observability/readme.md

```diff
@@ -9,9 +9,64 @@ Building observable systems enables development teams at CSE to measure how well
+- Ensure a downstream system mal-function won't cause massive repetitive logs being stored.
+- Don't reinvent the wheel, use existing tools to collect and analyse the data.
+- Ensure personal identifiable information policies and restrictions are followed.
+
+#### Having problems identifying what to log?
+
+**At application startup**:
+
+- Unrecoverable errors from startup
+- Warnings if application still runnable, but not as expected (i.e. not providing blob connection string, thus resorting to local files. Another example is if there's a need to fail back to a secondary service or a known good state, because it didn’t get an answer from a primary dependency.
+- Information about the service’s state at startup (build #, configs loaded, etc.)
+
+**Per incoming request**:
+
+- Basic information for each incoming request: scrubbed url, any user/tenant/request dimensions, response code returned, request-to-response latency, payload size, record counts, etc. (whatever you need to learn something from the aggregate data)
+- Warning for any unexpected exceptions, caught only at the top controller/interceptor and logged with or alongside the request info, with stack trace. Return a 500. This code doesn’t know what happened.
+
+**Per outgoing request**:
+
+- Basic information for each outgoing request: scrubbed url, any user/tenant/request dimensions, response code returned, request-to-response latency, payload sizes, record counts returned, etc. Report perceived availability and latency of dependencies and including slicing/clustering data that could help with later analysis.
+
+## Tracing
+
+Produces the information required to observe series of correlated operations in a distributed system. Once collected they show the path, measurements and faults in a end to end transaction.
+
+- Ensure that at least key business transactions are traced.
+- Include in each trace necessary information to identify software releases (i.e. service name, version). This is important to correlate deployments and system degradation.
+- Ensure dependencies are included in trace (databases, I/O).
+- If costs are a concern use sampling, avoiding throwing away errors, unexpected behaviour and critical information.
+- Don't reinvent the wheel, use existing tools to collect and analyse the data.
+- Ensure personal identifiable information policies and restrictions are followed.
+
+## Metrics
+
+Metrics provide a near real-time indication of how the system is running. As opposed to logs and traces, the amount of data collected using metrics remains constant as the system load increases.
+
+- Measurements that need to be alerted upon are good candidates for metrics: queue length, retries, latency, requests/sec, cpu, memory. These metrics can be use to automatically scale components.
+- Use metrics to collect business success measurements: visits, sold items, session time, games played, jobs done. Ask stakeholders what they need.
```

Comment:
Use metrics to collect business success measurements. Example - visits, sold items, session time, games played, jobs done. Ask stakeholders what they need